### ISSUE TITLE
Potential security vulnerabilities in the hdf5 library. Can you help upgrade to patch versions?

### ISSUE THREAD

**Issue by HelenParr** [unibas-gravis/scalismo-native#6]
Hi, @Ghazi-Bouabene , @madsendennis , I'd like to report a vulnerability issue in **ch.unibas.cs.gravisscalismo-native-all:4.0.1**.
### Issue Description
**ch.unibas.cs.gravisscalismo-native-all:4.0.1** directly or transitively depends on ***203*** C libraries (.so) cross many platforms(such as x86-64, x86, arm64, armhf). However, I noticed that one C library is vulnerable, containing the following CVEs:

`libjhdf5.so` from C project **hdf5(version:1.8.10)** exposed ***14*** vulnerabilities:
[CVE-2020-10811](https://nvd.nist.gov/vuln/detail/CVE-2020-10811), [CVE-2020-10812](https://nvd.nist.gov/vuln/detail/CVE-2020-10812), [CVE-2020-10810](https://nvd.nist.gov/vuln/detail/CVE-2020-10810), [CVE-2020-10809](https://nvd.nist.gov/vuln/detail/CVE-2020-10809), [CVE-2019-8396](https://nvd.nist.gov/vuln/detail/CVE-2019-8396), [CVE-2018-17437](https://nvd.nist.gov/vuln/detail/CVE-2018-17437), [CVE-2018-17432](https://nvd.nist.gov/vuln/detail/CVE-2018-17432), [CVE-2018-17433](https://nvd.nist.gov/vuln/detail/CVE-2018-17433), [CVE-2018-17434](https://nvd.nist.gov/vuln/detail/CVE-2018-17434), [CVE-2018-17438](https://nvd.nist.gov/vuln/detail/CVE-2018-17438), [CVE-2018-17436](https://nvd.nist.gov/vuln/detail/CVE-2018-17436), [CVE-2018-17233](https://nvd.nist.gov/vuln/detail/CVE-2018-17233), [CVE-2018-17234](https://nvd.nist.gov/vuln/detail/CVE-2018-17234), [CVE-2018-17237](https://nvd.nist.gov/vuln/detail/CVE-2018-17237)

Furthermore, the vulnerable methods in the vulnerable shared libraries can be actually invoked by Java code. For instance, following call chains can reach the vulnerable method(C code) H5AC_unpin_entry() in file `H5AC.c` reported by [CVE-2020-10810](https://nvd.nist.gov/vuln/detail/CVE-2020-10810).
```
call chains-----
H5Literate_by_name()->H5G_iterate()->H5G__obj_iterate()->H5G__dense_iterate()->H5HF_open()->H5HF_close()->H5HF_space_close()->H5FS_close()->H5FS_decr()->H5AC_unpin_entry()
```

### Suggested Vulnerability Patch Versions
***hdf5*** has fixed the vulnerabilities in versions ***>=1.12.1***

Java build tools cannot report vulnerable C libraries, which may induce potential security issues to many downstream Java projects.
Could you please upgrade the above shared libraries to their patch versions?

Thanks for your help~
Best regards,

**Comment by marcelluethi** [comment]
Thanks for reporting this. 

You may have noticed that the native-libs project is big, rather complicated and also quite difficult to deploy. This is the reason we were really slow in updating the native libraries. In the meantime we have made an attempt to disentangle and simplify the native libraries. We have split them up into several, independent libraries
- [hdf5nativelibs](https://github.com/unibas-gravis/hdf5nativelibs)
- [vtkjavanatives](https://github.com/marcelluethi/vtk-java-natives/)
- [niftijio](https://github.com/unibas-gravis/scalismo-niftijio)

We should easily be able to update the hdf5 libraries to a recent version within this new structure. The new version of Scalismo, which should be released within the next few days, will make use of these new libraries. Would upgrading to the latest scalismo version be possible for you? There should not be any breaking changes compared to scalismo 0.90. 

Best regards, 
Marcel

**Comment by marcelluethi** [comment]
@madsendennis @Andreas-Forster @Ghazi-Bouabene  @HelenParr 

I looked a bit more deeply into the issue and it turns out that, while possible, is takes quite some work to do it properly. As scalismo uses hdf5 through the object package, which in turn is part of the hdfviewer, we cannot only replace the hdf5 libraries. Rather, we also need to upgrade to the latest version of the object package, which in turn comes with additional dependencies that are a bit messy to incorporate. 

Using the hdf5 library from java is not pleasant at all. As we are using hdf5 only as a storage format and do not make use of its advanced features, it might be best to look for an alternative that is easier to maintain. One possibility would be to switch to [JHDF](https://github.com/jamesmudd/jhdf), which is a pure java library. As this library can only read and not write files, we would have to find an alternative file format for writing our models. I am not an expert when it comes to file formats for storing matrix data. However, at a first glance it seems that for example Bson [bson](https://bsonspec.org/) would do what we want. 

As such a change would have wider reaching consequences, I would like to open it to discussion. 
In particular, I think we should discuss the following questions:
- How critical are the reported vulnerabilities? How quickly would we need to act? 
- Do we need to support writing hdf5 also in the future, or would it be sufficient to support reading the hdf5 format
- What file format should we use if we decided to replace our standard format for saving model
- Who would be willing to help investigating this issue and possible new implementations?

**Comment by madsendennis** [comment]
It should definitely be addressed, but also not sure how fast (days, weeks, months). 

I still think that the .h5 format is a good format to store the models. I briefly looked into a few other formats and tried some loading/saving experiments in python.  
I converted a ~18mb .h5 to: .h5, .json, .bson, .npz. 

Below is average save time / load time / file size over 1000 runs: 
.h5: 37ms, 103us, 17.7mb
.npz: 50ms, 73us, 17.7mb
.bson: 7.2s, 2.95s, 54.6mb
.json 3.05, 1.21s, 101.8mb

Definitely, a lot of optimization to do as e.g. the json and bson had to convert the numpy arrays that the data was stored in to python lists. And the read/write was with the standard file read/write in python after serializing the data. 

I also found the bsdf format: http://bsdf.io/index.html, but it doesn't seem to be actively maintained, and there is no scala or java version implemented. 

Long story short, the best alternative as I see it would be .npz which is supported in multiple languages - but it lacks the functionality e.g. attributes from .h5 which should instead be dictionary entries. So I would be in favour of keeping the .h5 format. 

It also sounds like the jhdf library might support the writing of .h5 in the future.

Not sure how much time I can set aside for implementation - but will definitely help to test.

**Comment by Andreas-Forster** [comment]
I think we should look for a replacement for reading the files soon. It 
is definitively good that the issue was raised on the mailing list, in 
order that users of the lib are aware to not load random h5 files from 
somewhere. But if we can make it more safe by just adding a lib offering 
the reading, I think this is worth it.

Regarding a different file format, there I have no real opinion. Maybe 
it is even good to have a different, simpler one. Regarding the loading 
times and storage I am not scared from neither of the options. If one 
needs performance, it doesn't need a huge investment to save and load 
the data that we have in our h5 files.

Crucial I see, that we should support for quite some time the reading of 
the legacy format.

Further, I would also argue to offer a legacy write method, as it exists 
already and should not be much of a security problem. Could also be at a 
hidden spot of our library ;-)  This would offer companies that 
certified the h5 format to still use the old format, on their own risk.

In short:
- I would add reading through JHDF and keep the writer in.
- Then we have time to decide on the next better, near perfect file format.

I would also be willing to help to make this happen. Testing as well as 
helping to develop a PR or for discussion.


Am 5/3/2022 um 3:43 PM schrieb Dennis Madsen:

**Comment by Ghazi-Bouabene** [comment]
Similar to Andreas, I would also be in favor of:

**Short term:**

-  using JHDF for reading .h5 files (avoids the vulnerability)
-  keep using NCSA HDF5 for writing 

**Mid to Long term:**

* transition to a more flexible file format (has both technical and administrative advantages)
* keep support for reading legacy .h5 via JHDF 
* keep support for writing .h5 (while deperecating), either via JHDF if supported by then, or via NCSA  **<--**  this is important for users/companies that have slow/limited file format transition processes (due to certifications for example) 

Would also be happy to help here.

**Comment by marcelluethi** [comment]
Thanks for all the input. I opened a [PR](https://github.com/unibas-gravis/scalismo/pull/387) which proposes to change all reading routines to use jhdf. Writing is so far unchanged, but should not be problematic. 
The hope is that we can replace hdf5 in the not too distant future.

**Comment by marcelluethi** [comment]
The issue is fixed in scalismo version [0.91.1](https://github.com/unibas-gravis/scalismo/releases/tag/v0.91.1). 

I am closing the issue.
